0.x: Fix media not enabled by subscribers on publisher renegotiations

danxuliu opened the pull request:
Since 493632d1aaad when a subscriber is configured the `audio`, `video` and `data` flags are honoured. If they are disabled they will stay disabled, and if they are enabled then the result will depend on whether the publisher is sending them and the subscriber has offered to receive them. The value of the flags will be either [the value given in the configure request](https://github.com/meetecho/janus-gateway/blob/402189a1446fd53f4179126a7ce5ed3ba9860c6e/plugins/janus_videoroom.c#L7385-L7408) or the value they had already if it is not explicitly given.

[When the SDP is updated for a publisher the publisher sends a configure request to update all its subscribers](https://github.com/meetecho/janus-gateway/blob/402189a1446fd53f4179126a7ce5ed3ba9860c6e/plugins/janus_videoroom.c#L8244-L8268). Before 493632d1aaad if a publisher started to publish audio, video or data the audio, video or data was automatically enabled by the subscribers. However, after that change the audio, video or data is no longer automatically enabled; [if the audio, video or data was not being sent by the publisher when the subscriber was created the corresponding flag is disabled](https://github.com/meetecho/janus-gateway/blob/402189a1446fd53f4179126a7ce5ed3ba9860c6e/plugins/janus_videoroom.c#L6857-L6865), so when the publisher starts to send it it will stay disabled in the subscriber, as it was not possible to differentiate between explicitly disabled and implicitly disabled.

~~To solve that this pull request updates the `audio`, `video` or `data` flag of subscribers when subscribers are configured due to a change in the SDP of a publisher, thus restoring the behaviour previous to 493632d1aaad for that case (which was a side effect / regression, as the main target of that commit were updates of the subscriber itself*).~~

~~Note, however, that this implies that if a subscriber was explicitly configured to disable audio, video or data when the publisher was not sending it yet that setting will not be respected once the publisher starts to send it. The subscriber will need to be explicitly configured again to disable the audio, video or data. As the flags are booleans as far as I know there is currently no way to differentiate between the flags being explicitly disabled, or being implicitly disabled due to the publisher not sending the corresponding media (which could be used to automatically enable the media once available only in this case).~~

~~Nevertheless, the flags are updated only if the audio, video or data changed, to avoid overriding the subscriber flags when the audio, video or data did not change in the publisher SDP.~~

To solve that additional variables are introduced to keep track of whether the audio, video or data were explicitly configured or not and, if they were, with which value. This is now taken into account when a subscriber is updated, so the audio, video or data will be automatically enabled if the publisher starts to send them and they were not explicitly disabled by the subscriber.
    
Besides that, for consistency the configured value is now also taken into account when switching publishers for a subscriber; if the audio, video or data was explicitly disabled it will stay disabled after the switch.

*The main target of 493632d1aaad were [these two scenarios, both for subscribers](https://github.com/meetecho/janus-gateway/pull/2963#issuecomment-1113340998):
- Send `configure` request with `XXX: false` (where _XXX_ is `audio`, `video` or `data`), then send `configure` request with `update: true` -> Before 493632d1aaad _XXX_ was enabled, after 493632d1aaad it is still disabled
- Send `configure` request with `update: true` and `XXX: false` -> Before 493632d1aaad _XXX_ was still enabled (if it was enabled before), after 493632d1aaad it is disabled

januscla commented:
Thanks for your contribution, @danxuliu! Please make sure you sign our [CLA](https://cla.conf.meetecho.com/janus), as it's a required step before we can merge this.

lminiero commented:
I'm not sure reverting an explicit setting made before is a good idea, since that feels sneaky. IIRC the PR by @fancycode tried to address exactly that. As such, I think it's a better idea to check whether or not something was explicitly disabled, before implicitly enabling it.

danxuliu commented:
> I'm not sure reverting an explicit setting made before is a good idea, since that feels sneaky. IIRC the PR by @fancycode tried to address exactly that.

Yes and no :-) Back in the day @fancycode sent the fix to address some issues I mentioned to him, and those issues were about reverting the subscriber video when configuring the subscriber again. The fact that it also happened when publishers started to send media was just a side effect recently discovered.

Of course it is a good side effect if the subscriber media was explicitly disabled, but not a good one if it was just implicitly disabled and therefore should be automatically enabled once the publisher starts sending the media.

> As such, I think it's a better idea to check whether or not something was explicitly disabled, before implicitly enabling it.

I totally agree, but as mentioned in the description above as far as I know there is currently no way to differentiate between explicitly disabled or implicitly disabled. Would it be OK to add some additional flags to subscribers like `audio/video/data_was_configured` (just an example, please suggest the desired name :-) ) to keep track of that? I did not add them to keep the patch as simple as possible (even if I knew that I was trading off one side effect for the other), but I can try to do it if you want.

danxuliu commented:
I have dropped the previous commit and added the variables to track whether audio, video or data were configured or not and, if they were, whether they were enabled or not. Those variables are now taken into account when a subscriber is configured, so the media will be disabled if it was explicitly disabled (therefore the fix from 493632d is respected in all cases), and if it was explicitly enabled or not configured then it will depend on whether the publisher is sending the media and the subscriber offered to receive it (so the media will be automatically enabled if the subscriber was not configured and the publisher starts sending it).

A couple of questions:
- If the subscriber is switched to another publisher I kept the configured values, so if the media was disabled it will still be disabled after switching the publisher. Should they be reset instead (as it could be seen as a new subscriber)?
- Should the `XXX_configured` and `XXX_enabled` values be exposed [when querying the subscriber session](https://github.com/meetecho/janus-gateway/blob/402189a1446fd53f4179126a7ce5ed3ba9860c6e/plugins/janus_videoroom.c#L3004-L3011)?

Thanks!

lminiero commented:
> * f the subscriber is switched to another publisher I kept the configured values, so if the media was disabled it will still be disabled after switching the publisher. Should they be reset instead (as it could be seen as a new subscriber)?

I think it makes sense to preserve what was asked before even when switching, as it may be a subscription preference that's independent of who they're subscribed to (e.g. "I know I won't have enough bw for video so don't send me any").

> * Should the `XXX_configured` and `XXX_enabled` values be exposed [when querying the subscriber session](https://github.com/meetecho/janus-gateway/blob/402189a1446fd53f4179126a7ce5ed3ba9860c6e/plugins/janus_videoroom.c#L3004-L3011)?

In general it is indeed a good idea to add as much information as possible there, as `handle_info` requests are a common way to debug issues: as such, having access to what those values are set to can help figure out why something's not working (and people may have incorrectly used a feature or forgotten they'd enabled it).

danxuliu commented:
Sorry, I have noticed that I was not respecting the code style regarding white spaces for `if` statements. I have fixed that and force pushed again.

lminiero commented:
This looks good to me! Have you tested this in your setup and for your requirements after these last few changes? If so, I think it's good enough for me to merge.

danxuliu commented:
Sorry for the delay.

I have tested the following scenarios:
- Update an already disabled subscriber:
  - Publisher sending audio and video
  - Subscriber offering to receive audio and video
  - `configure` request for subscriber with `video: false`
  - `configure` request with `update: true`
  - Result: subscriber video still disabled
- Update a subscriber also disabling it
  - Publisher sending audio and video
  - Subscriber offering to receive audio and video
  - `configure` request for subscriber with `update: true` and `video: false`
  - Result: subscriber video disabled
- Update a publisher to start sending video
  - Publisher sending audio
  - Subscriber offering to receive audio and video
  - `configure` request for publisher with an updated SDP to send video
  - Result: subscriber video enabled
- Update a publisher to start sending video when the subscriber has video disabled
  - Publisher sending audio
  - Subscriber offering to receive audio and video
  - `configure` request for subscriber with `video: false`
  - `configure` request for publisher with an updated SDP to send video
  - Result: subscriber video disabled
- Update a publisher to start sending video when the subscriber has video disabled from an update
  - Publisher sending audio
  - Subscriber offering to receive audio and video
  - `configure` request for subscriber with `update: true` and `video: false`
  - `configure` request for publisher with an updated SDP to send video
  - Result: subscriber video disabled
- Querying the session (through Janus admin UI)

Unfortunately I could not test switching a subscriber to a different publisher. Sorry about that.

lminiero commented:
Thanks, this looks like a comprehensive set of checks to me! Merging then :v:
